runtime/buffer.cc

```diff
@@ -343,4 +343,40 @@ void fill(const raw_buffer& dst, const void* value) {
   fill(dst_base, dims, dst.elem_size, value, rank - 1);
 }
 
+namespace internal {
+
+void make_for_each_contiguous_slice_dims(const raw_buffer& buf, for_each_contiguous_slice_dim* dims) {
```

Comment:
This is pretty cool (and much more threadsafe)... and will likely make it easier to implement the multibuffer version on top of as well.

How does it compare on the benchmarks with the previous implementation?

Reply:
It's ~20% faster than the previous for tiny buffers.

I think it's the other helpers `for_each_tile` and `for_each_slice` that have the mutability problem. `for_each_contiguous_slice` didn't/doesn't need to mutate the buffer (because it doesn't call the callback with a buffer object).